src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -322,13 +384,99 @@ void remove_virtual_functionst::get_functions(
 
   // iterate over all children, transitively
   std::set<irep_idt> visited;
-  get_child_functions_rec(
+  get_child_functions(
     class_id,
     root_function.symbol_expr,
     component_name,
     functions,
     visited);
 
+  const auto find_empty = [&](const dispatch_table_entryt &entry) {
+    return entry.symbol_expr == symbol_exprt();
+  };
+  bool empty_found =
+    std::any_of(functions.begin(), functions.end(), find_empty);
+
+  if(empty_found && !functions.empty())
+  {
+    std::cout << "problematic dispatch table\n";
+    for(const auto &f : functions)
+      std::cout << f.class_id << ":" << f.symbol_expr.pretty() << " ";
+    std::cout << std::endl;
+
+    // with a non-empty virtual dispatch table but only empty function
+    // expressions, we could not find a class that provides the required
+    // method. Therefore we try finding it in the set of parent classes of a
+    // class in the dispatch table.
+
+    std::map<irep_idt, symbol_exprt> new_entries;
+    for(const auto &f : functions)
+    {
+      if(f.symbol_expr != symbol_exprt())
+        continue;
+
+      // search through functions of the parent classes
+      // const auto &parents = class_hierarchy.class_map[f.class_id].parents;
+      // for(const auto &p : parents)
+      // {
+        dispatch_table_entriest entries;
+      //   if(!class_hierarchy.is_abstract_class(p))
+      //   {
+          get_parent_functions(
+            f.class_id,
+            root_function.symbol_expr,
+            component_name,
+            entries,
+            visited);
+
+          // if valid entries are found, add them to the new dispatch table
+          const auto it =
+            std::find_if_not(entries.begin(), entries.end(), find_empty);
+          if(it != entries.end())
+            new_entries[f.class_id] = it->symbol_expr;
+          else
+            std::cout << "nothing found for " << f.class_id << std::endl;
+      //   }
+      // }
+    }
+
+    // collect sets of children classes for each class in the new additional
```

Comment:
I think this is incorrect. If we have an interface `Interface` that prototypes a method `foo`, and we have types `A`, `B` and `C`, such that (writing `<:` for subtyping) `A <: C, B <: C, A <: Interface`, and `C` and `B` define concrete definitions of `foo`, then a call `((Interface)something).foo()` may target an `A` object, and thus call `C.foo`, but may *not* target a `B` since that doesn't inherit `Interface`. This loop would appear to gather `B` as a candidate callee because it is a child of `C`.

In short, it should suffice to try to resolve a concrete function call at each child node where the method is abstract, without need for this further search over *child* types.

Reply:
@smowton thanks for your comments, I now use resolve-concrete-function which does this in a better way. For each child, I check for such an implementation if get_child_functions was not successful in finding one.

Reply:
I think it should check for one *even if* get-child-functions finds one. Example:

C extends B extends A. B implements Interface. C and A provide implementations of f. Then a call to ((Interface)x).f() should consider B.f() aka A.f() as a target, even though its child C provides an implementation.